contrib/lkt/language/parser.py

```diff
@@ -2270,21 +2270,36 @@ def check_correctness_pre():
             rd.has_error | et.has_error,
             No(T.SemanticResult.array),
 
-            If(
-                # We have a call to an object (either __call__ able object or
-                # function type object).
-                rd.is_null,
-                No(T.SemanticResult.array),
-
-                # Generic instantiation errors
-                rd.result_ref.cast(T.GenericDecl).then(
-                    lambda generic_decl: Entity.infer_actuals(
-                        generic_decl,
-                        Entity.expected_type
-                    ).then(lambda iir: iir.error.then(lambda e: e.singleton))
-                )
-            )._or(
-                # Arity errors
+            # First, check for generic instantiation errors. Before this, we
+            # cannot call ``Entity.called_decl``, because it assumes that
+            # generic instantiation errors have been checked beforehand.
+            rd.result_ref.cast(T.GenericDecl).then(
+                lambda generic_decl: Entity.infer_actuals(
+                    generic_decl,
+                    Entity.expected_type
+                ).then(lambda iir: iir.error.then(lambda e: e.singleton)),
+
+            )
+
+            # Handle errors for when we have a call to an object with no
+            # declaration, so we need to rely on the object_type, and the
+            # object type is non callable.
```

Comment:
I have trouble understanding this comment: what’s an object with no declaration? Do you mean when we have a call that we can’t resolve to a declaration?

Reply:
yes, amending